opendut-util/opendut-auth/src/confidential/blocking_client.rs

```diff
@@ -0,0 +1,226 @@
+                Ok(_) => {
+                    return Ok(());
+                }
+                Err(cause) => {
+                    error = Some(cause);
+                    tokio::time::sleep(Duration::from_millis(10000)).await;
+                    continue;
+                }
+            };
+        }
+        Err(ConfidentialClientError::KeycloakConnection { message: String::from("Could not connect to keycloak"), cause: error.unwrap() })
+    }
+
+    fn update_storage_token(response: &BasicTokenResponse, state: &mut RwLockWriteGuard<Option<TokenStorage>>) -> Result<Token, AuthError> {
+        let access_token = response.access_token().clone();
+        let expires_in = match response.expires_in() {
+            None => {
+                return Err(AuthError::ExpirationFieldMissing { message: "No expires_in in response.".to_string() });
+            }
+            Some(expiry_duration) => { Utc::now().naive_utc() + expiry_duration }
+        };
+        let _token_storage = state.insert(TokenStorage {
+            access_token,
+            expires_in,
+        });
+        Ok(Token { value: response.access_token().secret().to_string() })
+    }
+
+    fn fetch_token(&self) -> Result<Token, AuthError> {
+        let response = self.inner.exchange_client_credentials()
+            .add_scopes(self.config.scopes.clone())
+            .request(|request| { self.reqwest_client.sync_http_client(request) })
+            .map_err(|error| {
+                AuthError::FailedToGetToken {
+                    message: "Fetching authentication token failed!".to_string(),
+                    cause: WrappedRequestTokenError(error),
+                }
+            })?;
+
+        let mut state = self.state.write().expect("Failed to rewrite the token."); //TODO
+        
+        Self::update_storage_token(&response, &mut state)?;
+        Ok(Token { value: response.access_token().secret().to_string() })
+    }
+
+    pub fn get_token(&self) -> Result<Token, AuthError> {
+        let token_storage = self.state.read().unwrap().clone();
+        let access_token = match token_storage {
+            None => {
+                self.fetch_token()?
+            }
+            Some(token) => {
+                if Utc::now().naive_utc().lt(&token.expires_in) {
+                    Token { value: token.access_token.secret().to_string() }
+                } else {
+                    self.fetch_token()?
+                }
+            }
+        };
+        Ok(access_token)
+    }
+}
+
+#[cfg(test)]
+mod auth_tests {
+    use chrono::Utc;
+    use googletest::assert_that;
+    use googletest::matchers::gt;
+    use oauth2::{ClientId, ClientSecret, TokenResponse};
+    use url::Url;
+    use opendut_util_core::project;
+    use crate::confidential::config::ConfidentialClientConfigData;
+    use crate::confidential::pem::read_pem_from_file_path;
+    use crate::confidential::blocking_reqwest_client;
+
+    #[test]
+    fn test_get_token_example() {
```

Comment:
This integration test makes the CI fail. It requires a running keycloak instance. It should be ignored here.